### PR TITLE
Test word extraction changes

### PR DESCRIPTION
**Pull request by leohemsted**
add a new way to parse the address block of precompiled letters when we rewrite it.

But don't actually use this data for now - just log if it's different to the old way so we can gauge what impact we'd have if we changed our processes

---- 

My letter: image

What happens to it on production after we rewrite the address block in arial in the correct place: 
image

(ignore the square block for the ff in maffs for now - i was testing ligatures when i noticed this)

### CODE REVIEW

File: tests/test_precompiled.py [hunk 1]
```diff
@@ -724,6 +725,23 @@ def test_extract_address_block():
     )
 
 
+def test_extract_address_block_handles_unusal_y_coordinates(client, caplog):
```
Comment (whpearson) [inline tests/test_precompiled.py:728]: Typo unus**u**al

File: app/precompiled.py [hunk 2]
```diff
@@ -709,11 +709,29 @@ def _extract_text_from_page(page, rect):
...
     group = groupby(mywords, key=itemgetter(3))
     extracted_text = []
-    for _y1, gwords in group:
+    for _y2, gwords in group:
         extracted_text.append(" ".join(w[4] for w in gwords))
+
+    if _get_lines_by_paragraph_and_line(mywords) != extracted_text:
+        # grouping by paragraph ended up different to grouping by y2. lets just log for now. we might want to swap over
+        # in the future but without knowing how much it changes we cant be sure
+        current_app.logger.info("Address extraction differend between y2 and paragraph/line")
```
Comment (whpearson) [inline app/precompiled.py:733]: Typo differen**d**

**Review by whpearson — APPROVED** [review]

File: app/precompiled.py [hunk 3]
```diff
@@ -709,12 +710,25 @@ def _extract_text_from_page(page, rect):
...
+    extracted_text = "\n".join(extracted_text)
+
+    if _get_address_from_get_textwords() != extracted_text:
+        # grouping by paragraph ended up different to grouping by y2. lets just log for now. we might want to swap over
+        # in the future but without knowing how much it changes we cant be sure
+        current_app.logger.info("Address extraction different between y2 and get_text")
+
+    # normalizing to NFKD replaces characters with compatibility mode equivalents - including replacing
+    # ligatures like ﬀ with ff
+    return unicodedata.normalize("NFKD", "\n".join(extracted_text))
```
Comment (whpearson) [inline app/precompiled.py:731]: Actually this shouldn't be joined on "\n" as we have already done that on 722

Comment (leohemsted) [reply app/precompiled.py:731]: oops yes, bad rebase

**Review by whpearson — APPROVED** [review]

File: app/precompiled.py [hunk 4]
```diff
@@ -709,12 +709,23 @@ def _extract_text_from_page(page, rect):
...
     mywords = [w for w in words if fitz.Rect(w[:4]).intersects(rect)]
+
+    def _get_address_from_get_textwords():
+        return page.get_text(clip=rect).strip()
+
     mywords.sort(key=itemgetter(-3, -2, -1))
     group = groupby(mywords, key=itemgetter(3))
     extracted_text = []
-    for _y1, gwords in group:
+    for _y2, gwords in group:
```
Comment (CrystalPea) [inline app/precompiled.py:720]: why are we changing `_y1` to `_y2` here?

File: app/precompiled.py [hunk 5]
```diff
@@ -709,12 +709,23 @@ def _extract_text_from_page(page, rect):
...
         extracted_text.append(" ".join(w[4] for w in gwords))
-    return "\n".join(extracted_text)
+    extracted_text = "\n".join(extracted_text)
+
+    if _get_address_from_get_textwords() != extracted_text:
+        # grouping by paragraph ended up different to grouping by y2. lets just log for now. we might want to swap over
+        # in the future but without knowing how much it changes we cant be sure
+        current_app.logger.info("Address extraction different between y2 and get_text")
```
Comment (CrystalPea) [inline app/precompiled.py:727]: will we be able to tell for which letter it was, to see what the difference was? I vaguely remember we did some work to connect the whole request better in the logs?

File: app/precompiled.py (hunk 4 above)
Comment (leohemsted) [reply app/precompiled.py:720]: it more accurately describes the actual variable we're grouping on  from the tuple (x1, y1, x2, y2, word, para_num, line_num, word_num)

File: app/precompiled.py (hunk 5 above)
Comment (leohemsted) [reply app/precompiled.py:727]: it's possible, but a little clunky. you can see the request ID on the logs in template-preview, and work backwards from there

[admin sanitise flow](https://kibana.logit.io/s/9423a789-282c-4113-908d-0be3b1bc9d1d/goto/740770eda421d4a3823ecd6512093e45)
[api async flow](https://kibana.logit.io/s/9423a789-282c-4113-908d-0be3b1bc9d1d/goto/7ab2dc601886ccd32cc8fdec11aeda1d)

both of these will lead back to a notification ID or an upload ID which you can use to find the actual file and see what's up

File: app/precompiled.py (hunk 4 above)
Comment (CrystalPea) [reply app/precompiled.py:720]: ok, should we maybe explain in the commit that we are changing it - so it doesn't seem random?

**Review by whpearson — APPROVED** [review]

**Pull request merged**